[DependencyInjection] More bullet-proof expression evaluation

Wirone opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | n/a
| License       | MIT

As asked [here](https://x.com/_Codito_/status/1900523413204013096), we faced weird issue recently. I was able to reproduce it with minimal DI setup in `CheckTypeDeclarationsPassTest`.

Wirone committed:
Issue reproducer

See: https://x.com/_Codito_/status/1900523413204013096

Wirone committed:
More bullet-proof expression evaluation

In some scenarios evaluating expression can lead to TypeError (unresolved DI params being strings, not expected integers etc).

OskarStark commented on src/Symfony/Component/DependencyInjection/Tests/Fixtures/CheckTypeDeclarationsPass/ExpectsIntegerArgument.php:
```diff
@@ -0,0 +1,19 @@
+        return new self($foo);
+    }
+
+    public function instance(): self {
```
```suggestion
    public function instance(): self
    {
```

Wirone replied:
This should be caught by CI (Fabbot), please don't waste your precious time for being Human Linter 🙈.

(thanks for the review anyway 🍻)

stof replied:
fabbot ignores the `Fixtures` folder because we have many fixtures that are actually about generated code (used for comparisons for expectations) or about cases for which we want to explicitly test things that don't match the core coding styles for some reasons.

Wirone committed:
Fix styling

Wirone committed:
PHP 8.1 compatibility

`composer.json`'s requirement for PHP is `PHP >= 8.1`, so even though GH Actions does not contain PHP 8.1 in the matrix (which is wrong IMHO), let's stick with `readonly` on property level.

Wirone commented:
I've additionally fixed PHP 8.1 compatibility (`readonly` on fixture class level, I have this as a default in IDE and just today saw the red underline in files 😉). What's interesting is why [previous CI run](https://github.com/symfony/symfony/actions/runs/13859220017?pr=59976) did not run tests on PHP 8.1, and [latest one](https://github.com/symfony/symfony/actions/runs/13893375973?pr=59976) did 😅.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Tests/Fixtures/CheckTypeDeclarationsPass/ExpectsIntegerArgument.php:
```diff
@@ -0,0 +1,20 @@
+<?php
+
+declare(strict_types=1);
```
to be replaced by the licence header (same below)

keradus replied:
also in fixture file?
if so, we could run the Fixer to apply this for other fixture files

nicolas-grekas replied:
the thing is we'd like the licence header in fixtures (because why not), but then, no need to enforce other rules I'd say (at least we'll get many false-positives if we do)

stof replied:
and PHP-CS-Fixer only allows to exclude files entirely, not to override the config being applied to some files to apply only some of the rules.

Wirone replied:
Yeah, we definitely need to introduce skipping rules like ECS has, I like it a lot.

But you can also have a separate ruleset for fixtures and run Fixer twice, pointing to the config file.

nicolas-grekas reviewed (commented):
So, the error you describe on Twitter happens before env vars are unresolved, while running the expression.
I feel like the solution that's proposed at the moment is way to much. What if the error is a parse error or anything else that'd better be caught at linting time?
Sorry I don't have a proposal, that's just the question that comes up at the moment :)

Wirone commented:
> I feel like the solution that's proposed at the moment is way to much. What if the error is a parse error or anything else that'd better be caught at linting time?

@nicolas-grekas I've been thinking about your question last 2 days and my initial opinion did not change - `lint:container` is IMHO _not_ a place for spotting syntax errors. These should be caught in a separate job (`php -l` or parallel lint), the last moment for it could be `cache:clear`. I think `lint:container` should focus only on DI definition and skip all the errors related to initialising services. Ideally it shouldn't initialise them in the first place, but I get that in order to match signature's expected type with other service's type or expression' return type it has to be done. Anyway, I think my fix is valid, as it aligns with the comment that is right there:

> If a service from the expression cannot be fetched from the container, we skip the validation.

In our case it's runtime issue, but for other scenarios it can be actually language-level issue - none of them should break the DI linting (maybe flag for this would be a good idea, so you could run `lint:container --stop-on-error` or something like that).

nicolas-grekas commented:
What about checking the error message and ignoring only the ones we know want to ignore?
I don't agree with you about linting not being the place to spot parse errors: any opportunity to spot mistakes as easily as possible is a win. False-positives are what is bad, and I would care about them one by one.

nicolas-grekas commented:
Up to follow my last suggestion @Wirone?
